pkg/client/oci/oci.go

```diff
@@ -0,0 +1,91 @@
+package oci
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/jetstack/version-checker/pkg/api"
+	"github.com/jetstack/version-checker/pkg/version/semver"
+)
+
+type Client struct{}
+
+func New() (*Client, error) {
+	return &Client{}, nil
+}
+
+func (c *Client) Name() string {
+	return "oci"
+}
+
+func (c *Client) Tags(ctx context.Context, host, repo, image string) ([]api.ImageTag, error) {
+	src := fmt.Sprintf("%s/%s/%s", host, repo, image)
+	rpo, err := name.NewRepository(src)
+
+	if err != nil {
+		return []api.ImageTag{}, err
+	}
+
+	bareTags, err := remote.List(rpo, remote.WithContext(ctx))
+
+	if err != nil {
+		return []api.ImageTag{}, err
+	}
+
+	sort.SliceStable(bareTags, func(i, j int) bool {
+		return semver.Parse(bareTags[i]).LessThan(semver.Parse(bareTags[j]))
+	})
+
+	var tags []api.ImageTag
+	for i, t := range bareTags {
+		tag := api.ImageTag{Tag: t}
+
+		// Only fetch metadata for the highest version
+		// Could be very slow otherwise if there are a lot of tags, and
+		// we only really care about the highest version.
+		if i == len(bareTags)-1 {
+			img, err := name.ParseReference(fmt.Sprintf("%s:%s", src, t))
+			if err != nil {
+				continue
+			}
+			image, err := remote.Image(img, remote.WithContext(ctx))
```

Comment:
I believe `Image` will resolve manifest lists (multi-arch images) to a platform-specific manifest. The default is `linux/amd64`. Is this the desired behaviour?

Looking at the selfhosted client, I think it will return the `tag.SHA` and `tag.Tag` for the manifest list, it won't resolve down to a particular platform.

Based on that, I think this should be a `remote.Get` and then a check on `desc.MediaType` to figure out whether it's an image or an index. If it's an image then we can call `remote.Image` and extract the arch, OS and created time.

You can see here in the cosign codebase how they perform a similar check: https://github.com/sigstore/cosign/blob/main/pkg/oci/remote/remote.go#L71-L95.

Reply:
hm, yes - good call.

Reply:
I think this isn't relevant now that we're not looking to retrieve metadata in this first pass.